src/units/drawer.cpp

```diff
@@ -462,6 +469,29 @@ void unit_drawer::redraw_unit(const unit& u) const
 		halo_man.set_location(ac.unit_halo_, halo_x, halo_y);
 	}
 
+	const std::vector<std::string> halos_abilities = u.halo_abilities();
+	ac.size_vec_ = halos_abilities.size();
+	unsigned int n = std::max(ac.size_vec_, ac.size_vec_ref_);
+	for(unsigned i=0 ; i<n; ++i){
+		bool has_abil_halo = ac.abil_halos_[i] && ac.abil_halos_[i]->valid();
+		const std::string& halo_ab = !halos_abilities.empty() && (ac.size_vec_ > i) ? halos_abilities.at(i) : "";
+		if(!has_abil_halo && !halo_ab.empty()) {
+			ac.abil_halos_[i] = halo_man.add(
+				halo_x, halo_y,
+				halo_ab + u.TC_image_mods(),
+				map_location(-1, -1)
+			);
+		}
+		if(has_abil_halo && (halo_ab.empty() || (ac.abil_halos_ref_[i] != halo_ab))) {
+			halo_man.remove(ac.abil_halos_[i]);
+			ac.abil_halos_[i].reset();
+		} else if(has_abil_halo) {
+			halo_man.set_location(ac.abil_halos_[i], halo_x, halo_y);
+		}
+		ac.abil_halos_ref_[i] = halo_ab;
+	}
+	ac.size_vec_ref_ = ac.size_vec_;
```

Comment:
If I'm reading this correctly, this seems to be using `std::map` as a way of getting a `std::vector` that doesn't complain about using out-of-bounds indices, and it'll one-by-one remove and re-add halos if the size or order of the vector changes.

It feels like the wrong algorithm to do this, but am I misunderstanding what it's meant to do?

Reply:
I tried to do without map variables by using vector instead and opting for more global checking, but in this case I have overprinting problems and even after this first problem was solved, the set_location function was not working properly and I had the halo which struggled to follow the unit in its movement. This algorithm may not be ideal, but it's the only one that works perfectly

I'll keep it here for now

Reply:
@stevecotton algorythm changed for vector